### PR TITLE
Try to fix dereferencing the truth data

### PR DESCRIPTION
**Pull request by reunanen**

### CODE REVIEW

File: dlib/cuda/cuda_dlib.h [hunk 1]
```diff
@@ -592,10 +592,12 @@ namespace dlib
...
-                    // const_label_iterator& t = *truth;
-                    DLIB_ASSERT(t.nr() == subnetwork_output.nr());
-                    DLIB_ASSERT(t.nc() == subnetwork_output.nc());
+                    const auto& t = *truth;
+                    DLIB_ASSERT(t.size() == subnetwork_output.k());
+                    for (const auto& tk : t) {
+                        DLIB_ASSERT(tk.nr() == subnetwork_output.nr());
+                        DLIB_ASSERT(tk.nc() == subnetwork_output.nc());
+                    }
                     memcpy(buf + i*bytes_per_plane, &t, bytes_per_plane);
```
Comment (reunanen) [inline dlib/cuda/cuda_dlib.h]: Note that I didn't check (yet) if this is still correct.

Comment (reunanen) [reply dlib/cuda/cuda_dlib.h]: It doesn't look like it is. Will try to fix...

Comment (reunanen) [reply dlib/cuda/cuda_dlib.h]: Should be better now, I think.

**Comment by arrufat** [comment]
It seems to work on my machine, now. I am trying with a custom dataset to estimate landmarks using this loss function.
Using the CPU version:
```
step#: 0     learning rate: 0.1   average loss: 0            steps without apparent progress: 0
step#: 179   learning rate: 0.1   average loss: 0.0193513    steps without apparent progress: 29
step#: 365   learning rate: 0.1   average loss: 0.0025315    steps without apparent progress: 39
step#: 550   learning rate: 0.1   average loss: 0.00227839   steps without apparent progress: 122
step#: 734   learning rate: 0.1   average loss: 0.002129     steps without apparent progress: 53
```
Using the GPU version:
```
step#: 0     learning rate: 0.1   average loss: 0            steps without apparent progress: 0
step#: 171   learning rate: 0.1   average loss: 2.6403e-08   steps without apparent progress: 0
step#: 354   learning rate: 0.1   average loss: 3.26668e-09  steps without apparent progress: 25
step#: 541   learning rate: 0.1   average loss: 2.95653e-09  steps without apparent progress: 25
step#: 726   learning rate: 0.1   average loss: 2.84906e-09  steps without apparent progress: 123
```

So, I might have made some mistake in the GPU implementation.

Thee speed seems a little bit slower on the GPU version (since each line is printed every 40 seconds and the number of steps is slightly lower).

I will recheck that. Thank you so much for helping me out with this.
I am thinking about providing an example on how to use this loss to estimate keypoints, but I don't know if @davisking would like to have it on the official repo.
I've seen many people ask on the dlib issues how to train a model to estimate landmarks using the GPU. With this loss you can get surprisingly accurate results.
I took the main idea from here: https://arxiv.org/abs/1804.06208

**Comment by reunanen** [comment]
Ok. Yeah, I only checked superficially with `test/dnn.cpp`. It appeared to give similar results for the CPU and GPU implementations, but maybe it doesn't cover your situation. Are you able to expand the test so that the implementations differ like this? Or if you can just fix it directly, that works for me as well. :)

**Comment by arrufat** [comment]
I know how to reproduce it:
If you set `num_channels` to 10, the loss will be in the order of `1e-5` after a few epochs on the GPU, but around 0.07 on the CPU version.
The final error values would be similar, so I might be dividing by the number of channels twice somewhere.

**Comment by reunanen** [comment]
Right, you are multiplying by `scale` twice, aren't you?
(In `_cuda_compute_loss_mean_squared_per_channel_and_pixel`, but also in `compute_loss_mean_squared_per_channel_and_pixel::do_work`?)

**Comment by reunanen** [comment]
Note that the other CUDA kernels only multiply gradient elements by `scale`, and not the loss.

Guessing it's more efficient to multiply only once, i.e. in `do_work`.

**Comment by arrufat** [comment]
Yes, I was looking at that, too. Thanks I will fix it now :) and reduce the number of epochs and increase the number of channels in the test :)

Thank you so much!

**Comment by reunanen** [comment]
You are very welcome!
